assertion error with invalid SQL

### To reproduce

found via fuzzing:
```sql
SELECT fun(col BETWEEN (0 AND 1, col), col;
```

totally non-sensical SQL, but `AssertionError` is still not acceptable:
```
2025-12-14T17:52:35.738916Z C i.q.c.h.p.JsonQueryProcessorState [135291470075] internal error [ex=
java.lang.AssertionError: opStack is empty at ','
	at io.questdb.griffin.ExpressionParser.parseExpr(ExpressionParser.java:373)
	at io.questdb.griffin.SqlParser.expr(SqlParser.java:4292)
	at io.questdb.griffin.SqlParser.expr(SqlParser.java:4303)
	at io.questdb.griffin.SqlParser.parseSelectClause(SqlParser.java:3024)
	at io.questdb.griffin.SqlParser.parseDml0(SqlParser.java:1990)
	at io.questdb.griffin.SqlParser.parseDml(SqlParser.java:1880)
	at io.questdb.griffin.SqlParser.parseSelect(SqlParser.java:2972)
	at io.questdb.griffin.SqlParser.parse(SqlParser.java:4329)
	at io.questdb.griffin.SqlCompilerImpl.compileExecutionModel(SqlCompilerImpl.java:2564)
[...]
```

### QuestDB version:

c2343aad

### OS, in case of Docker specify Docker and the Host OS:

NA

### File System, in case of Docker specify Host File System:

NA

### Full Name:

Jaromir Hamala

### Affiliation:

QuestDB

### Have you followed Linux, MacOs kernel configuration steps to increase Maximum open files and Maximum virtual memory areas limit?

- [x] Yes, I have

### Additional context

_No response_

--- a/core/src/main/java/io/questdb/griffin/ExpressionParser.java
+++ b/core/src/main/java/io/questdb/griffin/ExpressionParser.java
@@ -1063,7 +1063,7 @@ public class ExpressionParser {
             int betweenAndCount = 0;
             int caseCount = 0;
             int argStackDepth = 0;
-            int betweenStartCaseCount = 0;
+            int betweenStartScopeDepth = 0;
             savedScopeStackBottom = scopeStack.getBottom();
             scopeStack.setBottom(scopeStack.sizeRaw());
             boolean parsedDeclaration = false;
@@ -1528,7 +1528,8 @@ public class ExpressionParser {
                                 processDefaultBranch = true;
                             }
                         } else if (SqlKeywords.isAndKeyword(tok)) {
-                            if (caseCount == betweenStartCaseCount && betweenCount > betweenAndCount) {
+                            if (scopeStack.size() == betweenStartScopeDepth
+                                    && betweenCount > betweenAndCount) {
                                 betweenAndCount++;
                                 thisBranch = BRANCH_BETWEEN_END;
                                 argStackDepth = popAndOpStack(listener, argStackDepth, prevBranch);
@@ -1603,7 +1604,7 @@ public class ExpressionParser {
                                 throw SqlException.$(lastPos, "between statements cannot be nested");
                             }
                             betweenCount++;
-                            betweenStartCaseCount = caseCount;
+                            betweenStartScopeDepth = scopeStack.size();
                         }
                         processDefaultBranch = true;
                         break;

--- a/core/src/test/java/io/questdb/test/griffin/SqlParserTest.java
+++ b/core/src/test/java/io/questdb/test/griffin/SqlParserTest.java
@@ -14746,6 +14746,51 @@ public class SqlParserTest extends AbstractSqlParserTest {
         );
     }
 
+    @Test
+    public void testBetweenAndSeparatorRespectsScopeDepth() throws Exception {
+        // AND inside a sub-expression that opens a new scope must NOT be claimed as
+        // BETWEEN's separator; only an AND at the same scope depth as BETWEEN is the
+        // separator. Covers every scope kind on scopeStack: PAREN, BRACKET, ARRAY,
+        // CASE, CAST/CAST_AS.
+
+        // PAREN
+        assertQuery(
+                "select-virtual 1 between (0 and 2, 3) column from (long_sequence(1))",
+                "SELECT 1 BETWEEN (0 AND 2) AND 3"
+        );
+
+        // ARRAY constructor
+        assertQuery(
+                "select-virtual 1 between (ARRAY[0 and 1], 3) column from (long_sequence(1))",
+                "SELECT 1 BETWEEN ARRAY[0 AND 1] AND 3"
+        );
+
+        // BRACKET (array indexing)
+        assertQuery(
+                "select-virtual 1 between (ARRAY[10, 20, 30][1 and 1], 3) column from (long_sequence(1))",
+                "SELECT 1 BETWEEN (ARRAY[10, 20, 30])[1 AND 1] AND 3"
+        );
+
+        // CASE
+        assertQuery(
+                "select-virtual 1 between (case when 1 = 1 and 2 = 2 then 0 else 0 end, 3) column from (long_sequence(1))",
+                "SELECT 1 BETWEEN CASE WHEN 1=1 AND 2=2 THEN 0 ELSE 0 END AND 3"
+        );
+
+        // CAST (and CAST_AS)
+        assertQuery(
+                "select-virtual 1 between ((0 and 1)::INT, 3) column from (long_sequence(1))",
+                "SELECT 1 BETWEEN CAST(0 AND 1 AS INT) AND 3"
+        );
+    }
+
+    @Test
+    public void testBetweenIssue6534() throws Exception {
+        // https://github.com/questdb/questdb/issues/6534
+        // Malformed SQL must throw SqlException, not AssertionError.
+        assertSyntaxError("SELECT fun(col BETWEEN (0 AND 1, col), col;", 42, "dangling literal");
+    }
+
     private void assertCreateTable(String expected, String ddl, TableModel... tableModels) throws SqlException {
         assertModel(expected, ddl, ExecutionModel.CREATE_TABLE, tableModels);
     }

